### DISCUSSION TITLE
150x image processing speedup with `zune-image`

### DISCUSSION THREAD

**Discussion by vlad-salone** [etemesi254/zune-image#181] [Show and tell]
While building my image optimization tool, I experimented with various image processing libraries. My initial attempt was homemade, but it proved buggy and hard to maintain. Switching to the `image` crate improved stability and performance.

However, I recently discovered `zune-image`, and it's a game changer! By leveraging zune-image's pipeline, I've achieved a significant performance leap. For single images, processing speeds are now up to 15 times faster compared to the previous version.

```
Benchmark 1: rimage-0.11.0-next.0
  Time (mean ± σ):      48.6 ms ±   1.3 ms    [User: 43.9 ms, System: 8.4 ms]
  Range (min … max):    46.7 ms …  53.3 ms    59 runs

Benchmark 2: rimage-0.10.3
  Time (mean ± σ):     691.6 ms ±   5.2 ms    [User: 709.3 ms, System: 10.1 ms]
  Range (min … max):   687.1 ms … 705.0 ms    10 runs

Benchmark 3: squoosh-cli
  Time (mean ± σ):     760.6 ms ±  29.7 ms    [User: 2155.7 ms, System: 291.2 ms]
  Range (min … max):   738.0 ms … 829.5 ms    10 runs

Summary
  rimage-0.11.0-next.0 ran
   14.23 ± 0.39 times faster than rimage-0.10.3
   15.65 ± 0.74 times faster than squoosh-cli
```

The real magic happens with batch processing. When optimizing around 15 images simultaneously on an 8-core system, zune-image delivers a staggering 150x speed increase. This translates to significant time savings, especially when dealing with large image sets.

```
Benchmark 1: rimage-0.11.0-next.0
  Time (mean ± σ):     177.2 ms ±   4.6 ms    [User: 999.5 ms, System: 38.7 ms]
  Range (min … max):   169.3 ms … 184.7 ms    17 runs

Benchmark 2: rimage-0.10.3
  Time (mean ± σ):     26.399 s ±  0.336 s    [User: 29.653 s, System: 0.089 s]
  Range (min … max):   26.045 s … 27.297 s    10 runs

Summary
  rimage-0.11.0-next.0 ran
  148.99 ± 4.31 times faster than rimage-0.10.3
```

My next steps is to optimize custom encoders and operations for even greater efficiency. In the latest pre-release of `rimage`, I've already builded custom encoders, decoders for AVIF and WebP formats, and MozJpeg with OxiPNG codecs. Additionally, I'm utilizing the `fast_image_resize` crate for incredibly fast resizing and `libimagequant` for quantization routines. With full release, all of this will be available for `zune-image` users.

**Comment by etemesi254** [comment]
Lessgoo, really nice encouragement here.

I designed the library to be fast and extendable, really nice to see it play that way. The api may be clumsy sometimes, but really nice to see it work this well, hope there weren't any huge hurdles.
